packages/sdk/src/TemplateProvider.ts

```diff
@@ -6,51 +6,91 @@ import * as fs from 'fs'
+        const templateDirectory = this.LGTemplateDirectory()
+        if (templateDirectory === null) {
             return null
         }
 
-        let templateString = JSON.stringify(template)
-        let argumentNames = this.GetArgumentNames(templateString)
-
-        // Substitute argument values
-        for (let argumentName of argumentNames) {
-            let renderedActionArgument = templateArguments.find(a => a.parameter == argumentName)
-            if (renderedActionArgument) {
-                templateString = templateString.replace(new RegExp(`{{${argumentName}}}`, 'g'), renderedActionArgument.value ?? '')
-            }
+        const renderingParameters = {}
+        // Copy parameters defined in the action to our rendering parameters.
+        for (const templateArgument of templateArguments) {
+            renderingParameters[templateArgument.parameter] = templateArgument.value
         }
-        templateString = this.RemoveEmptyArguments(templateString)
-        return JSON.parse(templateString)
+        // Copy entities in memory to our rendering parameters.
+        entityDisplayValues.forEach((value: string, key: string) => { renderingParameters[key] = value })
+
+        // Load the template.  Currently, we assume that each lg file only has one template.
+        const lgFilename = templateDirectory + "//" + templateName + ".lg"
+        const engine = new TemplateEngine().addFile(lgFilename)
+        const output = engine.evaluateTemplate(engine.templates[0].name, renderingParameters)
+        return ActivityFactory.createActivity(output)
     }
 
     public static GetTemplates(): Template[] {
-        let templates: Template[] = []
-        let files = this.GetTemplatesNames()
-        for (let file of files) {
-            let fileContent = this.GetTemplate(file)
-
-            // Clear submit check (will the set by extracting template variables)
-            this.hasSubmitError = false
-            let tvs = this.UniqueTemplateVariables(fileContent)
+        const files = this.GetTemplatesNames()
+        if (files.length === 0) {
+            return []
+        }
+        const templateDirectory = this.LGTemplateDirectory()
+        if (templateDirectory === null) {
+            throw new Error("Could not find valid template directory")
+        }
+        const templates: Template[] = []
+        for (const file of files) {
+            const fileName = path.join(templateDirectory, `${file}.lg`)
+            const engine = new TemplateEngine().addFile(fileName)
+            let templateBody = ''
+            if (file.includes('AdaptiveCard')) {
+                templateBody = engine.templates[1].body
+                templateBody = templateBody.replace("- ```\r\n", "").replace("```", "")
+                for (let symbol of engine.templates[1].parameters) {
+                    let sourceText = '@{' + symbol + '}'
+                    let replaceText = '{{' + symbol + '}}'
```

Comment:
This looks suspicious.  I thought the point of using LG was to adopt their templating language instead of using custom mustasch syntax.  Does the library not handle this?

Reply:
TBH, I find this chunk of code confusing also.  I don't know why or in what contexts an LG filename would include `AdaptiveCard` (seems fragile), or why the `@{}` syntax needs to be changed to `{{}}`.

There's also an issue that AFAICT, the `engine.templates[].parameters` property is always empty.  I've just been digging into this, and it appears to be why the UI for editing variables in the LG/CARD actions is never rendered.

@coldplaying42 - could you comment on these issues ?

Reply:
This is the hacking code to render the adaptive card for eye/preview icon.  The reason why  @{} syntax needs to be changed to {{}} is that in lg file we use @{} to refer to variables (e.g., "text": "@{question}"), while in json file we use {{}} to refer to variables (e.g.,  "text": "{{question}}"). I made this change to enable AdaptiveCardViewer process the templateString. 
https://github.com/microsoft/ConversationLearner-UI/blob/master/src/components/modals/AdaptiveCardViewer/AdaptiveCardViewer.tsx

However, this hacking code could only deal with the rendering of AdaptiveCard. In the future, AdaptiveCardViewer should be replaced with ActivityViewer, which can use some functions in https://github.com/microsoft/BotFramework-WebChat to render all the activities (e.g., thumbnail card, hero card, video card) that LG supports. And this chunk of code could be discarded. 

@tpmsr you could try the parameters-lg in https://github.com/coldplaying42/ConversationLearner-Samples/tree/zy_lg/lgs

Reply:
could you point me to the repo containing ActivityViewer please ?

Reply:
I think starting from https://github.com/microsoft/BotFramework-WebChat/tree/master/packages/bundle/src/adaptiveCards might be helpful. I am not familiar with BotFramework-WebChat. You could contact with BotFramework-WebChat team by Vishwac. They could help you to solve the eye/preview rendering problem.